backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/workpackage/impl/ShipmentLineForSOLogHandler.java

```diff
@@ -0,0 +1,192 @@
+				.flatrateTermId(FlatrateTermId.ofRepoId(flatrateTermRecord.getC_Flatrate_Term_ID()))
+				.build();
+
+		if (contractLogDAO.anyMatch(query))
+		{
+			return ExplainedOptional.emptyBecause("Contract Log already created for " + inOutLineRef);
+		}
+
+		final BPartnerId bPartnerId = BPartnerId.ofRepoId(flatrateTermRecord.getBill_BPartner_ID());
+
+		final UomId uomId = UomId.ofRepoId(inOutLineRecord.getC_UOM_ID());
+		final Quantity quantity = Quantitys.create(inOutLineRecord.getMovementQty(), uomId);
+		final WarehouseId warehouseId = WarehouseId.ofRepoId(inOutRecord.getM_Warehouse_ID());
+		final I_M_Warehouse warehouse = warehouseDAO.getById(warehouseId);
+
+		final ProductId productId = ProductId.ofRepoId(inOutLineRecord.getM_Product_ID());
+
+		final String description = msgBL.getMsg(MSG_INFO_SHIPMENT_SO_COMPLETED, ImmutableList.of(String.valueOf(productId.getRepoId()), quantity.toString()));
+
+		return ExplainedOptional.of(LogEntryCreateRequest.builder()
+											.contractId(createLogRequest.getContractId())
+											.productId(ProductId.ofRepoId(inOutLineRecord.getM_Product_ID()))
+											.referencedRecord(TableRecordReference.of(I_M_InOutLine.Table_Name, inOutLineRecord.getM_InOutLine_ID()))
+											.collectionPointBPartnerId(BPartnerId.ofRepoId(warehouse.getC_BPartner_ID()))
+											.producerBPartnerId(bPartnerId)
+											.invoicingBPartnerId(bPartnerId)
+											.warehouseId(WarehouseId.ofRepoId(inOutRecord.getM_Warehouse_ID()))
+											.documentType(LogEntryDocumentType.SHIPMENT)
+											.contractType(LogEntryContractType.MODULAR_CONTRACT)
+											.soTrx(SOTrx.SALES)
+											.processed(false)
+											.quantity(quantity)
+											.amount(null)
+											.transactionDate(LocalDateAndOrgId.ofTimestamp(inOutRecord.getMovementDate(), OrgId.ofRepoId(inOutLineRecord.getAD_Org_ID()), orgDAO::getTimeZone))
+											.year(createLogRequest.getModularContractSettings().getYearAndCalendarId().yearId())
+											.description(description)
+											.modularContractTypeId(createLogRequest.getTypeId())
+											.build());
+	}
+
+	@Override
+	public @NonNull ExplainedOptional<LogEntryReverseRequest> createLogEntryReverseRequest(
+			@NonNull final HandleLogsRequest<I_M_InOutLine> handleLogsRequest,
+			@NonNull final FlatrateTermId contractId)
+	{
+		final I_M_InOutLine inOutLineRecord = handleLogsRequest.getModel();
+
+		final TableRecordReference inOutLineRef = TableRecordReference.of(I_M_InOutLine.Table_Name, inOutLineRecord.getM_InOutLine_ID());
+
+		final Quantity quantity = contractLogDAO.retrieveQuantityFromExistingLog(ModularContractLogQuery.builder()
+																						 .flatrateTermId(contractId)
+																						 .referenceSet(TableRecordReferenceSet.of(inOutLineRef))
+																						 .build());
+
+		final String description = msgBL.getMsg(MSG_INFO_SHIPMENT_SO_REVERSED, ImmutableList.of(String.valueOf(inOutLineRecord.getM_Product_ID()), quantity.toString()));
```

Comment:
Same. Pls use baseLanguage as requested in TC and also change it in ``ShipmentLineForPOLogHandler`` pls

Reply:
done